text/0000-autoref-copy.md

```diff
@@ -0,0 +1,199 @@
+- Feature Name: autoref_copy
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC allows autoreferencing of `T` where `T: Copy`.
+
+Example:
+
+```rust
+[derive(Copy, Clone)]
+struct S(u8);
+
+fn by_val(_: S) {}
+fn by_ref(_: &S) {}
+fn by_mut_ref(_: &mut S) {}
+
+fn main() {
+    let s = S(5);
+    by_val(s);
+    by_ref(s); // This works after this RFC.
+    // by_mut_ref(s); // ERROR -- expected `&mut S`, found `S`
+}
+```
+
+# Motivation
+[motivation]: #motivation
+
+When working with `Copy` data, the distinction between borrowed and owned data
+is often unimportant. However, generic code often results in code which will
+only accept a particular variant of a `Copy` type (`T`, `&T`, `&&T`, ...).
+This is a frustration in many places:
+
+```rust
+// Comparisons:
+let v = vec![0, 1, 2, 3];
+v.iter().filter(|x| x > 1); // ERROR: expected &&_, found integral variable
+// These work:
+0 > 1;
+&0 > &1;
+// But these don't:
+&0 > 1;
+0 > &1;
+
+// Trait instantiations:
+let mut map: HashMap::new();
+map.insert(0, "Hello");
+map.insert(1, "world!");
+map[0]; // ERROR: expected &{integer}, found integral variable
+// or
+map.get(1); // ERROR: expected `&{integer}`, found integral variable
+
+// Numeric operators:
+// These work:
+&0 + &1;
+0 + &1;
+&0 + 1;
+// But these don't:
+&&0 + 1;
+&&0 + &1;
+&&0 + &&1;
```

Comment:
The RFC should note that the proposed coercion will not make the examples with binary operations compile. Or the examples should be removed, they don't really motivate this RFC.